qa/L0_response_cache/ensemble_cache_test.py

```diff
@@ -0,0 +1,296 @@
+            f"Failed inference recorded for ensemble model '{self.ensemble_model}'. Expected successful inference.",
+        )
+        self.assertIn(
+            "count",
+            ensemble_model_stats["cache_hit"],
+            f"No cache hit recorded for ensemble model '{self.ensemble_model}'. Expected exactly one cache hit.",
+        )
+        self.assertEqual(
+            actual_cache_hit_count,
+            expected_cache_hit_count,
+            f"Unexpected number of cache hits recorded for ensemble model '{self.ensemble_model}'. Expected exactly one cache hit.",
+        )
+
+    def test_ensemble_all_models_cache_enabled(self):
+        """
+        Test top level response caching when response cache enabled in
+        all the models.
+        Expected result: One cache hit in ensemble model stats. No cache hit in composing model stats.
+        """
+        self._update_config(
+            self.ensemble_config_file, RESPONSE_CACHE_PATTERN, RESPONSE_CACHE_CONFIG
+        )
+        self._update_config(
+            self.composing_config_file, RESPONSE_CACHE_PATTERN, RESPONSE_CACHE_CONFIG
+        )
+        self._run_inference_and_validate(self.ensemble_model)
+        ensemble_model_stats = self._get_model_statistics(self.ensemble_model)
+        composing_model_stats = self._get_model_statistics(self.composing_model)
+        expected_cache_hit_count = "1"
+        actual_cache_hit_count = ensemble_model_stats["cache_hit"]["count"]
+        self.assertIn(
+            "count",
+            ensemble_model_stats["success"],
+            f"Failed inference recorded for ensemble model '{self.ensemble_model}'. Expected successful inference.",
+        )
+        self.assertIn(
+            "count",
+            ensemble_model_stats["cache_hit"],
+            f"No cache hit recorded for ensemble model '{self.ensemble_model}'. Expected exactly one cache hit.",
+        )
+        self.assertNotIn(
+            "count",
+            composing_model_stats["cache_hit"],
+            f"Unexpected cache hit recorded for composing model '{self.composing_model}'. Expected top-level response in cache for ensemble model '{self.ensemble_model}'.",
+        )
+        self.assertEqual(
+            actual_cache_hit_count,
+            expected_cache_hit_count,
+            f"Unexpected number of cache hits recorded for ensemble model '{self.ensemble_model}'. Expected exactly one cache hit.",
+        )
+
+    def test_ensemble_composing_model_cache_enabled(self):
+        """
+        Test caching behavior when response cache enabled only in
+        composing model's config file.
+        Expected result: One cache hit in composing model stats. No cache related metric counts in
+        ensemble model stats.
+        """
+        self._update_config(
+            self.composing_config_file, RESPONSE_CACHE_PATTERN, RESPONSE_CACHE_CONFIG
+        )
+        self._run_inference_and_validate(self.composing_model)
```

Comment:
Should we be running inference on the ensemble model here, and validating that ensemble did inference but has no cache stats, and that composing model does have correct cache stats? Looks like we're doing inference on composing model directly here so not actually testing the ensemble flow.

Reply:
We are running inference on ensemble model only. The model parameter is only to verify baseline stats in run_inference_and_validate. For this testcase, ensemble model stats are going to be empty. So I passed model as a parameter to correctly verify corresponding model's stats, for this testcase it's composing model.